src/core/qgsauxiliarystorage.cpp

```diff
@@ -19,13 +19,129 @@
 #include "qgslogger.h"
 #include "qgsslconnect.h"
 #include "qgsproject.h"
+#include "qgspallabeling.h"
+#include "qgsdiagramrenderer.h"
 
 #include <QFile>
 
 const QString AS_JOINFIELD = "ASPK";
 const QString AS_EXTENSION = "qgd";
 const QString AS_JOINPREFIX = "auxiliary_storage_";
 
+QgsAuxiliaryField::QgsAuxiliaryField( const QgsPropertyDefinition &def )
+  : QgsField()
+  , mPropertyDefinition( def )
+{
+  init( def );
+}
+
+QgsAuxiliaryField::QgsAuxiliaryField( const QgsField &f )
+{
+  const QStringList parts = f.name().split( '_' );
+
+  if ( parts.size() <= 1 )
+    return;
+
+  const QString origin = parts[0];
+  const QString propertyName = parts[1];
+  QgsPropertyDefinition def;
+
+  if ( origin.compare( "pal", Qt::CaseInsensitive ) == 0 )
+  {
+    const QgsPropertiesDefinition props = QgsPalLayerSettings::propertyDefinitions();
+    Q_FOREACH ( const QgsPropertyDefinition p, props.values() )
+    {
+      if ( p.name().compare( propertyName, Qt::CaseInsensitive ) == 0 )
+      {
+        def = p;
+        break;
+      }
+    }
+  }
+  else if ( origin.compare( "diagram", Qt::CaseInsensitive ) == 0 )
+  {
+    const QgsPropertiesDefinition props = QgsDiagramLayerSettings::propertyDefinitions();
+    Q_FOREACH ( const QgsPropertyDefinition p, props.values() )
+    {
+      if ( p.name().compare( propertyName, Qt::CaseInsensitive ) == 0 )
+      {
+        def = p;
+        break;
+      }
+    }
+  }
+
+  if ( !def.name().isEmpty() )
+  {
+    init( def );
+    setTypeName( f.typeName() );
+    mPropertyDefinition = def;
+  }
+}
+
+void QgsAuxiliaryField::init( const QgsPropertyDefinition &def )
+{
+  if ( !def.name().isEmpty() )
+  {
+    QVariant::Type type;
+    int len( 0 ), precision( 0 );
+    switch ( def.dataType() )
+    {
+      case QgsPropertyDefinition::DataTypeString:
+        type = QVariant::String;
+        len = 50;
+        break;
+      case QgsPropertyDefinition::DataTypeNumeric:
+        type = QVariant::Double;
+        len = 0;
+        precision = 0;
+        break;
+      case QgsPropertyDefinition::DataTypeBoolean:
+        type = QVariant::Int; // sqlite does not have a bool type
+        break;
+      default:
```

Comment:
Remove this default - it's better to get the compiler warning when new types are added

Reply:
done